File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.kt
import java.util.HashMap

class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        val sLen = s.length
        if (sLen == 0) {
            return 0
        }
        val dic = HashMap<Char, Int>()
        var left = 0
        var right = 0
        var answer = 0

        while (right < sLen) {
            when {
                !dic.containsKey(s[right]) || dic[s[right]] == 0 -> {
                    dic[s[right]] = dic.getOrDefault(s[right], 0) + 1
                    right++
                }
                else -> {
                    dic[s[left]] = dic.getOrDefault(s[left], 0) - 1
                    left++
                }
            }
            answer = maxOf(answer, right - left)
        }
        return answer
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import defaultdict

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic = defaultdict(int)
        left = 0
        answer = 0
        for right, val in enumerate(s):
            if val in dic and dic[val] >= left:
                left = dic[val] + 1
            dic[val] = right
            answer = max(answer, right - left + 1)
        return answer
                
